template: adding toTime function to TemplateExpander

jrkt opened the pull request:
This PR adds a `formatTime` function in the TemplateExpander. As it stands now, the way to get the current time of a PromQL execution is to do something like this: `{{ query "time()" | first | value || humanizeTime }}`. This works great but only allows one format of time to be expressed. I'm proposing a simple function that does virtually the exact same thing as `humanizeTime` except that it allows you to pass in the format you expect to get back like this: `{{ query "time()" | first | value || formatTime "2006-01-02 15:04:05" }}`

dgl commented:
Thanks, I think it would be better to instead add a "toTime" function that takes a unix epoch time and can convert it to a `time.Time`, that way other methods are available. (Formatting would look something like `{{ (query "time()" | first | value | toTime).Format("...") }}`)

(This is similar to what is possible in alertmanager, e.g. https://github.com/prometheus/alertmanager/issues/1188#issuecomment-904624486)

jrkt commented:
I can do that.

jrkt commented:
@dgl done.

jrkt commented:
@roidelapluie I have broken out the pieces I could into a `floatToTime` function. I had to leave the `convertToFloat` in both functions as the converted float64 is used directly in `humanizeTimestamp`. Although, I don't know how that error would ever be hit. I have also added some tests. I am unable to add test cases for testing a float64 against the two functions `math.IsNaN` or `math.IsInf`. For `math.IsNan`, the function literally is `return f != f` and for `math.IsInf` the compiler fails because the float64 overflows. Lastly, I updated the documentation to include a "Time" section since now there are two functions directly related to time. Let me know if I missed anything.

jrkt commented:
@roidelapluie I have separated out the pieces of the functions that I could. I had to leave `convertToFloat` out of the shared function since the converted float is used in an error case in `humanizeTimestamp`. Everything else has been broken out though. I also added tests and updated the documentation. I added a new "Time" section to the documentation since there are now two functions directly related to time. Let me know if I missed anything.

jrkt commented:
@roidelapluie is there anything I missed here?

roidelapluie commented on docs/configuration/template_reference.md:
```diff
@@ -57,7 +57,13 @@ If functions are used in a pipeline, the pipeline value is passed as the last ar
 | humanizeDuration   | number or string | string  | Converts a duration in seconds to a more readable format. |
 | humanizePercentage | number or string | string  | Converts a ratio value to a fraction of 100. |
-| humanizeTimestamp  | number or string | string  | Converts a Unix timestamp in seconds to a more readable format. |
+
```
The list is not long enough for splitting I'd say.

jrkt commented:
@roidelapluie requests resolved.

roidelapluie commented:
Thanks!
